### PR TITLE
lsi: introduce the WD-1600 microassembler

### PR DESCRIPTION
**Pull request by 1801BM1**
The simple microassembler to compile the microcode sources
is written with the following features:

- follows the original semantics of DEC MICASM
- supports .radix 8 or 10
- supports local labels
- supports arithmetic expressions
- compiles multiple source files
- checks the translation locations for LSI-11
- portable, written in Phyton 3.x

Signed-off-by: 1801BM1@gmail.com

### CODE REVIEW

File: lsi/rom/tools/cp16mic.py [hunk 1]
```diff
@@ -0,0 +1,981 @@
...
+import os
+import re
+import sys
+import argparse
+
+MIN_LC = 0x0000
+MAX_LC = 0x0800
+
+T_SYM = 0
+T_REG  = 1
```
Comment (yshestakov) [inline lsi/rom/tools/cp16mic.py]: `pep8` will issue a warning on extra space before `=` there. In fact `pep8` generates a lot of warnings related to spaces in the code

File: lsi/rom/tools/cp16mic.py [hunk 2]
```diff
@@ -0,0 +1,981 @@
...
+	        'LRR'   : (1, T_SYM),
+	        'RSVC'  : (2, T_SYM)
+        }
+        return
+
+    def dir_title(self, opcode, word):
+        #
+        # Ignore .title for compatibility
+        #
+        return -1
```
Comment (yshestakov) [inline lsi/rom/tools/cp16mic.py:132]: Please convert comments in the header of function to docstrings like this:

```
    def dir_title(self, opcode, word):
        """
        Ignore .title for compatibility
        """
        return -1
```

File: lsi/rom/tools/cp16mic.py (hunk 1 above)
Comment (1801BM1) [reply lsi/rom/tools/cp16mic.py]: OK, will run PEP-8 either

File: lsi/rom/tools/cp16mic.py [hunk 3]
```diff
@@ -0,0 +1,981 @@
...
+
+    def dir_radix(self, opcode, word):
+        radx = self.eval_exp(opcode[1])
+        if radx not in (8, 10):
```
Comment (yshestakov) [inline lsi/rom/tools/cp16mic.py:140]: why not radix16 there?

Comment (1801BM1) [reply lsi/rom/tools/cp16mic.py:140]: radix-16 is ambigous
hard to distinguish valid names and digits: BAC0 - might be name or digit

File: lsi/rom/tools/cp16mic.py [hunk 4]
```diff
@@ -0,0 +1,981 @@
...
+        except ValueError:
+            raise SyntaxError("invalid integer expression '%s'" % token)
+    #
+    # Token generator wrapper, replaces the symbols with actual integers
+    #
+    def subst_symbol(self, string):
+        for token in string:
+            if token in wd16.EXP_ALL:
+                yield token
+            elif token[0] in '0123456789':
```
Comment (yshestakov) [inline lsi/rom/tools/cp16mic.py:520]: do you take into account value of `self.radx` == 8 or 10 there?

File: lsi/rom/tools/cp16mic.py (hunk 3 above)
Comment (yshestakov) [reply lsi/rom/tools/cp16mic.py:140]: I see

File: lsi/rom/tools/cp16mic.py (hunk 4 above)
Comment (yshestakov) [reply lsi/rom/tools/cp16mic.py:520]: BTW, `subst_number()` supports hexadecimal constants

File: lsi/rom/tools/cp16mic.py [hunk 5]
```diff
@@ -0,0 +1,981 @@
...
+    def close_file(self, file):
+        if file is not None:
+            file.close()
+        return
+
+    #
+    # Show final compilation statistics
+    #
+    def final_stat(self, flst):
+        msg = '\r\nErrors: %d\r\nWarnings: %d\r\n' % (self.ecnt, self.wcnt)
```
Comment (yshestakov) [inline lsi/rom/tools/cp16mic.py:615]: Assuming you have your code running on Windows and the output file is opened in "t" (text) mode, you don't need to add extra `\r` symbol there.  For Unix `\r` is not needed as well.

File: lsi/rom/tools/cp16mic.py [hunk 6]
```diff
@@ -0,0 +1,981 @@
...
+            if oc and self.ps == 2:
+                ext = self.symb.get(oc)
+                if ext is None or ext & ~0x7F:
+                    raise SyntaxError("invalid translation code '%s'" % oc)
+                word |= ext << 24
+        try:
+            word = obj[2](self, opcode, word)
+        except SyntaxError as err:
+            if obj[0] < 0:
+                raise SyntaxError(err)
```
Comment (yshestakov) [inline lsi/rom/tools/cp16mic.py]: In this case you will have 2 `SyntaxError` objects, where there embedded one contains "message string" as an argument, and outer one contains `err` as `SyntaxError` instance.
It is enough just to do  `raise err` here

File: lsi/rom/tools/cp16mic.py [hunk 7]
```diff
@@ -0,0 +1,981 @@
...
+    params = parser.parse_args()
+
+    try:
+        asm = wd16()
+        #
+        # Check all source files existance
+        # 
+        for src in params.src:
+            fsrc = asm.open_file(src, 'mic', 'r')
+            asm.close_file(fsrc)
```
Comment (yshestakov) [inline lsi/rom/tools/cp16mic.py:1009]: in POSIX you have special API `access()` to check that file exists w/o opening it. In Python you have `os.path.exists(filename)` function

File: lsi/rom/tools/cp16mic.py (hunk 4 above)
Comment (1801BM1) [reply lsi/rom/tools/cp16mic.py:520]: Valid identifier can't start with digits, so it is filtering out the potential digits. If radix is 8 then int() will raise exception later

Comment (1801BM1) [reply lsi/rom/tools/cp16mic.py:520]: > BTW, `subst_number()` supports hexadecimal constants
Yes:
- 0xnnn is always treated as hex
- 0bnnn  is always treated as bin
- nnnn. is always treated as dec
- 0nnn is always treated as oct
- nnnn is treated on radix settings

File: lsi/rom/tools/cp16mic.py (hunk 6 above)
Comment (1801BM1) [reply lsi/rom/tools/cp16mic.py]: nice to know, thanks

File: lsi/rom/tools/cp16mic.py (hunk 7 above)
Comment (1801BM1) [reply lsi/rom/tools/cp16mic.py:1009]: The objective is to check everything, including the access rights

**Pull request merged**